### PR TITLE
refactor(sqllogictests): port group by test to sqllogic

### PR DESCRIPTION
**Pull request by aprimadi**
# Which issue does this PR close?

Port group by tests from group_by.rs.

Work towards #4495 

# Rationale for this change

N/A

# What changes are included in this PR?

N/A

# Are these changes tested?

Yes

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @aprimadi  -- we really appreciate the help

**Comment by aprimadi** [comment]
Not sure if we should port [group_by_dictionary](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/group_by.rs#L745) test case @alamb. 

It seems that if we port it into sqllogic, the test would be longer. While doing it programmatically seems to be very concise.

**Comment by aprimadi** [comment]
This is probably enough for now, all the other tests use a one-time-use test data.

File: datafusion/core/tests/sqllogictests/src/setup.rs [hunk 1]
```diff
@@ -39,6 +41,95 @@ pub async fn register_aggregate_tables(ctx: &SessionContext) {
     register_test_data(ctx);
 }
 
+pub async fn register_group_tables(ctx: &SessionContext) {
+    register_time_test_tables(ctx);
+}
+
+/// Create and populate time32_s, time32_ms, time64_us, time64_ns tables
+fn register_time_test_tables(ctx: &SessionContext) {
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/src/setup.rs:49]: I don't think you need rust code to do this setup anymore

You can create arbitrary Arrow types that have no corresponding SQL types with the `arrow_cast` function -- https://arrow.apache.org/datafusion/user-guide/sql/data_types.html

For example, here is one way to do it:
https://github.com/apache/arrow-datafusion/blob/7759d96e892e7c159136af35ebc0def49ed3b3b1/datafusion/core/tests/sqllogictests/test_files/timestamps.slt#L113-L129

So you can make a value of `Time32(Seconds)` with something like

```sql
❯ select arrow_cast(5000::int, 'Time32(Second)');
+-------------+
| Int64(5000) |
+-------------+
| 01:23:20    |
+-------------+
```

Comment (aprimadi) [reply datafusion/core/tests/sqllogictests/src/setup.rs:49]: Alright, thank you for the feedback @alamb. Didn't know that. I'm done for the day, will probably work more on this tomorrow if nothing comes up.

**Comment by aprimadi** [comment]
@alamb not sure how to port [group_by_date_trunc](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/group_by.rs#L648) and [group_by_dictionary](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/group_by.rs#L745) without writing setup code. Do we want to migrate those also?

**Comment by jiangzhx** [comment]
> 

My idea is mainly based on these two test cases and what they cover.

1. [group_by_date_trunc](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/group_by.rs#L648) based on the method name, is to test date_trunc. I think we can use an existing test data and move to sqllogictests
3. [group_by_dictionary](https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/sql/group_by.rs#L745) based on the code, is testing DictionaryArray. In my opinion, it should not be migrated to sqllogictest.

**Comment by aprimadi** [comment]
I think both `group_by_date_trunc` and `group_by_dictionary` test should not be moved.

For `group_by_date_trunc` the data is procedurally generated and it also generates multiple partition file.

For `group_by_dictionary`, I totally agree with you @jiangzhx

IMO, the data setup should be as close to the code that test it except if it's being used multiple times.

**Comment by aprimadi** [comment]
Btw,, those are the only two tests from group_by.rs not being moved to sqllogic.

**Comment by alamb** [comment]
I really appreciate the help @aprimadi  ❤️

**Pull request merged**